core/src/main/java/com/scalar/db/transaction/consensuscommit/Snapshot.java

```diff
@@ -513,11 +469,116 @@ void toSerializable(DistributedStorage storage)
     parallelExecutor.validate(tasks, getId());
   }
 
-  private TableMetadata getTableMetadata(Scan scan) throws ExecutionException {
-    TransactionTableMetadata metadata = tableMetadataManager.getTransactionTableMetadata(scan);
+  private void validateScanResults(
+      DistributedStorage storage,
+      Scan scan,
+      Map<Key, TransactionResult> results,
+      boolean scanWithLimit)
+      throws ExecutionException, ValidationConflictException {
+    Scanner scanner = null;
+    try {
+      // Only get tx_id, tx_version and primary key columns because we use only them to compare
+      scan.clearProjections();
+      scan.withProjection(Attribute.ID).withProjection(Attribute.VERSION);
+      ScalarDbUtils.addProjectionsForKeys(scan, getTableMetadata(scan));
+
+      if (!scanWithLimit) {
+        scanner = storage.scan(scan);
+      } else {
+        // Get a scanner without the limit if the scan has a limit
+        scanner = storage.scan(Scan.newBuilder(scan).limit(0).build());
+      }
+
+      Iterator<Entry<Key, TransactionResult>> resultsIterator = results.entrySet().iterator();
+
+      Optional<Result> latestResult = scanner.one();
+
+      Entry<Key, TransactionResult> resultsEntry = null;
+      if (resultsIterator.hasNext()) {
+        resultsEntry = resultsIterator.next();
+      }
+
+      while (latestResult.isPresent() && resultsEntry != null) {
```

Comment:
Is the original limit not taken into account in this method?

Reply:
Right. We don't need to consider the original limit. We only need to consider the number of the scan result here.